lti_tool/views.py

```diff
@@ -0,0 +1,320 @@
+    access_mode = ACCESS.TEACHER
+
+    def get_resource_objects(self):
+        super().get_resource_objects()
+        user = self.request.user
+        profile = UserProfile.objects.get(user=user)
+        self.is_teacher = (Enrollment.objects
+            .filter(user_profile=profile,role=Enrollment.ENROLLMENT_ROLE.TEACHER).exists())
+
+    def get_common_objects(self):
+        super().get_common_objects()
+        self.parse_lti_session_params()
+        user = self.request.user
+        profile = UserProfile.objects.get(user=user)
+        self.teacher_enrollments = (Enrollment.objects.select_related("course_instance")
+            .filter(user_profile=profile, role=Enrollment.ENROLLMENT_ROLE.TEACHER))
+
+
+class LtiSelectContentView(BaseTemplateView, LtiSelectContentMixin):
+
+    template_name = "lti_tool/course_list.html"
+
+    def get_common_objects(self):
+        super().get_common_objects()
+        self.instances = [e.course_instance for e in self.teacher_enrollments]
+        self.note("instances")
+
+
+class LtiSelectCourseView(CourseInstanceBaseView, LtiSelectContentMixin):
+
+    template_name = "lti_tool/module_list.html"
+
+    def post(self, request, *args, **kwargs):
+        # Create gradebook items
+        ags = self.message_launch.get_ags()
+        lineitems = ags.get_lineitems()
+        exercises = BaseExercise.objects.filter(course_module__course_instance=self.instance)
+        for e in exercises:
+            if e.max_points > 0 and len([li for li in lineitems if li['tag'] == str(e.id)]) == 0:
+                li = LineItem()
+                li.set_tag(str(e.id)).set_score_maximum(e.max_points).set_label(str(e))
+                ags.find_or_create_lineitem(li)
+
+        # Send activity settings
+        deep_link = self.message_launch.get_deep_link()
+        resource = DeepLinkResource()
+        (resource.set_url(request.build_absolute_uri(reverse('lti-launch')))
+            .set_custom_params({
+                "course_slug": kwargs['course_slug'],
+                "instance_slug": kwargs['instance_slug']
+            })
+            .set_title("{}: {}".format(self.course.name, self.instance.instance_name)))
+        return HttpResponse(deep_link.output_response_form([resource]))
+
+    # Get list of modules in course
+    def get(self, request, *args, **kwargs):
+        self.modules = list(self.instance.course_modules.all())
+        self.note("modules")
+        return super().get(request, *args, **kwargs)
+    
+
+class LtiSelectModuleView(CourseModuleBaseView, LtiSelectContentMixin):
+
+    template_name = "lti_tool/exercise_list.html"
+
+    def post(self, request, *args, **kwargs):
+        # Create gradebook items
+        ags = self.message_launch.get_ags()
+        lineitems = ags.get_lineitems()
+        exercises = BaseExercise.objects.filter(course_module=self.module)
+        for e in exercises:
+            if e.max_points > 0 and len([li for li in lineitems if li['tag'] == str(e.id)]) == 0:
```

Comment:
The expression is supposed to check if the lineitem already exists in the LTI platform in order to avoid unnecessary requests - in testing these sometimes slowed things down a lot. But you're right, it's confusing. How about the following?

```suggestion
            lineitem_exists = [li for li in lineitems if li['tag'] == str(e.id)]
            if e.max_points > 0 and not lineitem_exists:
```

Reply:
That looks great! 

An alternative in the functional style could look like the following, but in Python it is not very pretty. Yours looks nicer!

```python
any(map(lambda li: li['tag'] == str(e.id), lineitems))
```